Add Isolatable<T> interface

eljobe opened the pull request:
Add `Isolatable<T>` abstraction

This change adds the interface and several implementations to allow `SnapshotValue` objects to be isolatable as long as they are running in the same JVM in which they were created.

This change goes on further to integration the usage of this new interface into the artifact transform code which to show how it can make it possible to run several of those transforms in parallel.

eljobe commented:
:x: DO NOT MERGE!

adammurdoch commented on subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/ValueSnapshotter.java:
```diff
@@ -129,6 +142,17 @@ private SerializedValueSnapshot serialize(Object value) {
     }
 
+    private ValueSnapshot wrap(Object value, ValueSnapshot possible) {
+        if (possible instanceof EnumValueSnapshot) {
```
We don't want to retain strong references to the enum values or the value's class when we only need a snapshot (rather than an isolatable), for example when we're snapshotting a task's properties. I would add another public method whose responsibility is to create an `Isolatable` and leave the behaviour of `snapshot()` as it was, and use this new method for artifact transforms registration.

eljobe replied:
Done

adammurdoch commented on subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/DefaultVariantTransformRegistration.java:
```diff
@@ -66,7 +67,13 @@
+        }
+        ArrayValueSnapshot paramsSnapshot = (ArrayValueSnapshot) snapshot;
+
+        this.transformer = new ArtifactTransformBackedTransformer(implementation, paramsSnapshot, instantiator);
```
we should add some kind of functional test to show that this actually works.

eljobe replied:
Added

eljobe commented:
Thanks for the feedback @adammurdoch. I'll get started on both these changes right away.

oehme commented on subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/ArrayValueSnapshot.java:
```diff
@@ -69,4 +71,19 @@ public boolean equals(Object obj) {
+            if (snapshot instanceof Isolatable) {
+                toReturn[i] = ((Isolatable) snapshot).isolate();
+            } else {
+                throw new IsolationException("Attempted to isolate an object which is not Isolatable.");
```
❌ The message should contain the offending object

eljobe replied:
Done

oehme commented on subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/MapValueSnapshot.java:
```diff
@@ -70,4 +73,17 @@ public boolean equals(Object obj) {
+    @Override
+    public Map isolate() {
+        Map map = new HashMap();
+        for (ValueSnapshot key : entries.keySet()) {
```
⭕️ Iterating over the entryset is more efficient.

eljobe replied:
Done

oehme commented on subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/SetValueSnapshot.java:
```diff
@@ -67,4 +72,17 @@ public boolean equals(Object obj) {
+
+    @Override
+    public Set isolate() {
+        Set set = new HashSet();
```
⭕️ I think we have to use LinkedHashSet here, as the original might have been linked as well.

eljobe replied:
Done

eljobe commented:
Okay. I think I've addressed all the comments from @adammurdoch and @oehme.

I'm not crazy about needing to introduce `ValueSnapshotStrategy` and `IsolatableValueSnapshotStrategy`, but I think it was the right call at this point in the development of the code.

Please take another look at your earliest convenience.

oehme commented on subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/MapValueSnapshot.java:
```diff
@@ -70,4 +79,21 @@ public boolean equals(Object obj) {
+
+    @Override
+    public Map isolate() {
+        Map map = new HashMap();
```
❌ LinkedHashMap

eljobe replied:
Done

oehme commented on subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformIsolationTest.groovy:
```diff
@@ -0,0 +1,194 @@
+        and:
+        outputContains("variants: [{artifactType=thirdCount}, {artifactType=thirdCount}]")
+        file("build/libs3").assertHasDescendants("test-1.3.jar.txt", "test2-2.3.jar.txt")
+        file("build/libs3/test-1.3.jar.txt").readLines() == ["2", "3", "4", "5", "6"]
```
If I read the setup correctly, all these transforms should have exactly the same output. The arguments should be isolated the first time the transform is actually needed, not when it is created. Is this something we plan to address now or later?

eljobe replied:
I *think* you aren't reading the setup correctly. But, maybe we should chat about this.

In between registering each of the three transforms, I mutate the state of `counter` with the `increment()` method. The first transform snapshots the state of the `counter` with value `0`. The second with `1` and the third with `2`.  So, each of the two files which are transformed with the same transform are producing the same outputs. For example, both `libs3/test-1.3.jar.txt` and `libs3/test2-2.3.jar.txt` end up with lines `["2", "3", "4", "5", "6"]` because they were both registered when the counter had the value `2`.  It's important that even though the transform itself is mutating the counter's state by calling `increment()` the transform that is run for the other file, doesn't see the mutation.

eljobe replied:
After discussion with @oehme, it is clear that we aren't 100% aligned on what we're trying to build.

@adammurdoch told me this morning in a slack conversation:
> we want to verify that
> 1. none of the transforms see changes made by the build logic after registration, and
> 2. none of the transforms see changes made by the other transforms

I think we are all agreed on part 2. But, @oehme makes a good case for NOT wanting part 1.

Essentially, snapshotting at transform eagerly at registration time means that the user can no longer configure the parameters to the transform after it's registered. But, that's exactly what they will sometimes want to do. For example, a plugin might register a transform that depends on an Android build flavor, and then later in the configuration, in the user's build script, they may configure the build flavor to use some custom dexer. We want to notice that change and perform the correct transform "in isolation()" only just before we execute the transform.

@adammurdoch, please respond whether or not you agree with @oehme's recommendation to defer both snapshotting AND isolation until just before applying the transform.

oehme replied:
The Android plugin does everything in an `afterEvaluate` block, so they'd be fine, but more generally I think it would be good to snapshot/isolate only when the transform is actually needed. That can be a follow-up story.

adammurdoch replied:
Let's make it a separate story and get this stuff into master. We can have the discussion about whether and if so how on that other story.

eljobe replied:
@adammurdoch, sounds like a good plan to me. Do you want to review the rest of the code as well, or are you implying review approval in your last comment here?

oehme commented on subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/DefaultVariantTransformRegistration.java:
```diff
@@ -55,18 +57,23 @@
         hasher.putHash(classLoaderHierarchyHasher.getClassLoaderHash(implementation.getClassLoader()));
 
         // TODO - should snapshot later?
```
This is the TODO I was referring to. I think it would be the right thing to snapshot lazily.
